### PR TITLE
Improving the disparity filter interface

### PR DESCRIPTION
**Pull request by sbokov**
Several changes to minimize the need for explicit parameter tuning when using with either StereoBM or StereoSGBM and to make the filter more easy to use in general.

### CODE REVIEW

File: modules/ximgproc/include/opencv2/ximgproc/disparity_filter.hpp [hunk 1]
```diff
@@ -117,10 +117,24 @@ class CV_EXPORTS_W DisparityWLSFilter : public DisparityFilter
...
-/** @brief Factory method, create instance of DisparityWLSFilter and execute the initialization routines.
+/** @brief Convenience factory method that creates an instance of DisparityWLSFilter and sets up all the relevant
+filter parameters automatically based on the matcher instance. Currently supports only StereoBM and StereoSGBM.
+
+@param matcher_left stereo matcher instance that will be used with the filter
+
+@param dst_matcher_right this matcher instance should be used to generate right disparity maps and pass them to the filter
+*/
+CV_EXPORTS_W
+Ptr<DisparityWLSFilter> createDisparityWLSFilter(Ptr<StereoMatcher> matcher_left, Ptr<StereoMatcher>& dst_matcher_right);
```
Comment (mshabunin) [inline modules/ximgproc/include/opencv2/ximgproc/disparity_filter.hpp]: I'd suggest to divide this method to:
```.cpp
createDisparityWLSFilter(left_matcher)
// and
setupRightMatcher(left_matcher) // or 'createRightMatcher'
```

File: modules/ximgproc/tutorials/disparity_filtering.markdown [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+processing on CPU.
+
+In this tutorial you will learn how to use the disparity map post-filtering to improve the results
+of StereoBM and StereoSGBM algorithms.
+
+Source Stereoscopic Image
+-------------------------
+
+![Left view](images/ambush_5_left.png)
+![Right view](images/ambush_5_right.png)
```
Comment (mshabunin) [inline modules/ximgproc/tutorials/disparity_filtering.markdown]: It is better to use JPEG images with medium or high compression to optimize page size for online view.

File: modules/ximgproc/tutorials/disparity_filtering.markdown [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+
+Source Code
+-----------
+
+@code{.cpp}
```
Comment (mshabunin) [inline modules/ximgproc/tutorials/disparity_filtering.markdown]: Let's use full `samples/disparity_filtering.cpp` contents instead:
```.cpp
@include ximgproc/samples/disparity_filtering.cpp
```

File: modules/ximgproc/tutorials/disparity_filtering.markdown [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+        imwrite(dst_im, disparity_map_visualization);
+
+    return 0;
+}
+@endcode
+
+Explanation
+-----------
+-#  **Load left and right views**
+    @code{.cpp}
```
Comment (mshabunin) [inline modules/ximgproc/tutorials/disparity_filtering.markdown]: Let's use snippets from original `disparity_filtering.cpp` example:
- First mark blocks in the example:
```.cpp
//! [some_block]
...
//! [some_block]
```
- Then use this block in the tutorial:
```.cpp
@snippet ximgproc/samples/disparity_filtering.cpp some_block
```

**Comment by mshabunin** [comment]
Looks good, thank you!
Please, squash all commits into one.

**Comment by sbokov** [comment]
@mshabunin I replaced the code in the tutorial with a link to GitHub and squashed all the commits

**Comment by mshabunin** [comment]
:+1:

**Comment by Arqu** [comment]
Is it possible that this change introduced this error when building:
`no member named 'MODE_SGBM_3WAY' in 'cv::StereoSGBM'` found in the file disparity_filtering.cpp from the samples folder?

**Comment by mshabunin** [comment]
@Arqu , please use latest master branch version of opencv repository when building latest version of opencv_contrib.

This mode has been added recently:
https://github.com/Itseez/opencv/blob/master/modules/calib3d/include/opencv2/calib3d.hpp#L1559

**Comment by Arqu** [comment]
Thanks, I'll rebuild it when I get some spare time, I've managed to skip the module in the build process for now. I probably didn't use the latest source as I've used the 3.0.0 branch instead of master.

**Pull request merged**